cpp/src/arrow/python/arrow_to_pandas.cc

```diff
@@ -1105,21 +1162,24 @@ using BlockMap = std::unordered_map<int, std::shared_ptr<PandasBlock>>;
 // * placement arrays as we go
 class DataFrameBlockCreator {
  public:
-  explicit DataFrameBlockCreator(const std::shared_ptr<Table>& table) : table_(table) {}
+  explicit DataFrameBlockCreator(PandasOptions options,
+                                 const std::shared_ptr<Table>& table, MemoryPool* pool)
+      : table_(table), options_(options), pool_(pool) {}
 
   Status Convert(int nthreads, PyObject** output) {
     column_types_.resize(table_->num_columns());
     column_block_placement_.resize(table_->num_columns());
     type_counts_.clear();
     blocks_.clear();
 
-    RETURN_NOT_OK(CreateBlocks());
+    RETURN_NOT_OK(InferOutputBlockTypes());
+    RETURN_NOT_OK(CreateTrivialBlocks());
```

Comment:
Take a look at what I did on this. I think this code can be incrementally improved to have fewer special cases. We should add more ASV benchmarks so that we can refactor without fear of altering performance / parallelism